stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -493,19 +493,32 @@ rdiv!(A::AbstractMatrix{T}, transD::Transpose{<:Any,<:Diagonal{T}}) where {T} =
 (\)(A::Union{QR,QRCompactWY,QRPivoted}, B::Diagonal) =
     invoke(\, Tuple{Union{QR,QRCompactWY,QRPivoted}, AbstractVecOrMat}, A, B)
 
-@inline function kron!(C::AbstractMatrix, A::Diagonal, B::Diagonal)
+
+function diag_kron!(C, A, B)
     valA = A.diag; nA = length(valA)
     valB = B.diag; nB = length(valB)
     nC = checksquare(C)
     @boundscheck nC == nA*nB ||
         throw(DimensionMismatch("expect C to be a $(nA*nB)x$(nA*nB) matrix, got size $(nC)x$(nC)"))
+    
     @inbounds for i = 1:nA, j = 1:nB
         idx = (i-1)*nB+j
         C[idx, idx] = valA[i] * valB[j]
     end
     return C
 end
 
+@inline function kron!(C::AbstractMatrix{T}, A::Diagonal, B::Diagonal) where T
```

Comment:
I'm not sure the `diag_kron!` dispatch is really necessary. `fill!(::Diagonal, ...)` only fills the diagonal vector. It's slightly redundant, but avoiding this operation leads to 10 lines of code.

Reply:
sure, I'll just use the fill! version then.